### DISCUSSION TITLE
Serving non-html content from non-api route

### DISCUSSION THREAD

**Discussion by Niveous** [vercel/next.js#12835] [Help] [answered]
I have posts at `/posts/[slug].tsx` and I want to return a plain text response from `posts/[slug]/raw.tsx`. Is there any way to do this? I want this to work with `next export` so I can't do it in an API route.

**Comment by brunocrosier** [comment]
I think you'd have to put your `raw.tsx` file in `./public/posts/[slug]/` 

However, you will probably need to manually do the `[slug]` bit

**Reply by Niveous** [reply]
Thanks. This seems like the only way to solve this. Will see if someone else can find a better way.

**Comment by leerob** [comment, marked as answer]
Could you share a little more about what you're building? Trying to understand the use case for a plain-text response with dynamic routing.

**Reply by Niveous** [reply]
I want to serve a plaintext (markdown really) version of each post from the post url + `/raw`.

**Reply by leerob** [reply]
So plaintext file, or Markdown? I think there's an important distinction there. If markdown, then I'd just throw all the content in a single `div` unstyled or similar.

**Reply by Niveous** [reply]
I want to serve the markdown as plain text. Similar to how Reddit serves json for each page: https://www.reddit.com/r/javascript.json

**Reply by leerob** [reply]
It gets a bit tricky since you're limited to `next export`. 

You could create a node script that runs during the build process and executes something similar to an API route.

**Reply by leerob** [reply]
For example, I run a node script to [export a sitemap](https://github.com/leerob/leerob.io/blob/master/next.config.js#L24) during the build process in `next.config.js`.

**Reply by Niveous** [reply]
> For example, I run a node script to [export a sitemap](https://github.com/leerob/leerob.io/blob/master/next.config.js#L24) during the build process in `next.config.js`.

That looks good. Does it only run the script on `next export`?

**Reply by leerob** [reply]
It runs when you build your application with `next build`.

**Reply by Niveous** [reply]
That should be fine. Thank you!

**Comment by secam** [comment]
There is an option to do res.write(plaintext) and then call res.end(), that would avoid rendering the page and return just the plaintext to the client.

**Reply by flavio20002** [reply]
It's explained here:

https://stackoverflow.com/questions/64936872/next-js-is-there-any-way-to-render-an-xml-file

```
import { GetServerSideProps } from 'next'
import React from 'react'

const Sitemap: React.FC = () => null

export const getServerSideProps: GetServerSideProps = async ({ res }) => {
  if (res) {
    res.setHeader('Content-Type', 'text/xml')
    res.write(`<?xml version="1.0" encoding="UTF-8"?>
    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">

    </urlset>`)
    res.end()
  }
  return {
    props: {},
  }
}

export default Sitemap

```

**Reply by stephenafamo** [reply]
This is great. Thank you.

**Reply by ignaciolarranaga** [reply]
Guys is there something similar to this with the app router?, or the recommendation is to keep using the page router?
I mean for example returning a site index for a sitemap.

**Reply by goodhoko** [reply]
This doesn't work with `next export` which OP explicitly asked for.
